src/libdrakvuf/linux-exports.c

```diff
@@ -0,0 +1,447 @@
+#define VM_WRITE	        0x00000002
+#define VM_EXEC		        0x00000004
+#define VM_SHARED	        0x00000008
+#define R_X86_64_GLOB_DAT	0x00000006
+
+addr_t process_sym2va(drakvuf_t drakvuf, drakvuf_trap_info_t* info, vmi_pid_t pid, const char* lib, const char* sym)
+{
+    vmi_instance_t vmi = drakvuf->vmi;
+
+    addr_t process_base = drakvuf_get_current_process(drakvuf, info);
+
+    if (!drakvuf_get_process_pid(drakvuf, process_base, &pid))
+        return -1;
+
+    addr_t mm_struct_address;
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_PID,
+        .addr = process_base + drakvuf->offsets[TASK_STRUCT_MMSTRUCT],
+        .pid = pid
+    };
+    if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &mm_struct_address))
+        return -1;
+
+    addr_t mmap;
+    ctx.addr = mm_struct_address + drakvuf->offsets[MM_STRUCT_MMAP];
+    if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &mmap))
+        return -1;
+
+    char* libname = "";
+    addr_t vm_next, nullp = 0;
+
+    addr_t text_segment_address = 0;
+    addr_t data_segment_address = 0;
+    addr_t text_segment_size = 0;
+    do
+    {
+        addr_t vm_start;
+        ctx.addr = mmap + drakvuf->offsets[VM_AREA_STRUCT_START];
+        if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &vm_start))
+            return -1;
+
+        addr_t vm_end;
+        ctx.addr = mmap + drakvuf->offsets[VM_AREA_STRUCT_END];
+        if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &vm_end))
+            return -1;
+
+        ctx.addr = mmap + drakvuf->offsets[VM_AREA_STRUCT_NEXT];
+        if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &vm_next))
+            return -1;
+
+        addr_t file_address;
+        ctx.addr = mmap + drakvuf->offsets[VM_AREA_STRUCT_FILE];
+        if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &file_address))
+            goto next;
+
+        addr_t path_dentry;
+        ctx.addr = file_address + drakvuf->offsets[FILE_PATH] + drakvuf->offsets[PATH_DENTRY];
+        if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &path_dentry))
+            goto next;
+
+        ctx.addr = path_dentry + drakvuf->offsets[DENTRY_D_NAME] + drakvuf->offsets[QSTR_NAME] + 16;
+        libname = vmi_read_str(vmi, &ctx);
+        PRINT_DEBUG("LIB NAME is: %s \n", libname);
+
+        addr_t pgoffset;
+        ctx.addr = mmap + drakvuf->offsets[VM_AREA_STRUCT_PGOFF];
+        if (VMI_FAILURE == vmi_read_64(vmi, &ctx, &pgoffset))
+            goto next;
+        pgoffset = pgoffset << PAGE_SHIFT;
+
+        addr_t vm_flags;
+        ctx.addr = mmap + drakvuf->offsets[VM_AREA_STRUCT_FLAGS];
+        if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &vm_flags))
+            goto next;
+
+        if (strncmp(libname, lib, strlen(lib)) == 0 && (vm_flags & VM_READ) && (vm_flags & VM_EXEC))
+        {
+            text_segment_address = vm_start;
+        }
+
+        if (strncmp(libname, lib, strlen(lib)) == 0 && (vm_flags & VM_READ) && !(vm_flags & VM_WRITE)&& !(vm_flags & VM_EXEC))
```

Comment:
You do the `strncmp(libname, lib, strlen(lib)) == 0 && (vm_flags & VM_READ)` check twice here, code smell